runtime/plugin/matchparen.vim

```diff
@@ -1,225 +1,297 @@
+  enddef
+
+else
+
+  # fall back on matchaddpos() otherwise
+  def Highlight()
+    w:matchparen = matchaddpos('MatchParen', [
+      [c_lnum, c_col - before],
+      [m_lnum, m_col]
+    ])
+  enddef
+
+endif
+
+def Toggle(enable = !exists('#matchparen')) #{{{2
+  if enable
+    Autocmds(true)
+    ParseMatchpairs()
+    UpdateHighlight()
   else
-    let timeout = exists("b:matchparen_timeout") ? b:matchparen_timeout : g:matchparen_timeout
+    Autocmds(false)
+    RemoveHighlight()
   endif
-  try
-    let [m_lnum, m_col] = searchpairpos(c, '', c2, s_flags, s_skip, stopline, timeout)
-  catch /E118/
-    " Can't use the timeout, restrict the stopline a bit more to avoid taking
-    " a long time on closed folds and long lines.
-    " The "viewable" variables give a range in which we can scroll while
-    " keeping the cursor at the same position.
-    " adjustedScrolloff accounts for very large numbers of scrolloff.
-    let adjustedScrolloff = min([&scrolloff, (line('w$') - line('w0')) / 2])
-    let bottom_viewable = min([line('$'), c_lnum + &lines - adjustedScrolloff - 2])
-    let top_viewable = max([1, c_lnum-&lines+adjustedScrolloff + 2])
-    " one of these stoplines will be adjusted below, but the current values are
-    " minimal boundaries within the current window
-    if i % 2 == 0
-      if has("byte_offset") && has("syntax_items") && &smc > 0
-	let stopbyte = min([line2byte("$"), line2byte(".") + col(".") + &smc * 2])
-	let stopline = min([bottom_viewable, byte2line(stopbyte)])
-      else
-	let stopline = min([bottom_viewable, c_lnum + 100])
-      endif
-      let stoplinebottom = stopline
+enddef
+
+def InStringOrComment(): bool #{{{2
+  # Expression which detects whether the current cursor position is in certain
+  # syntax types (string, comment, etc.); evaluated inside lambda passed as
+  # skip argument to searchpairpos().
+  return synstack('.', col('.'))
+    ->mapnew((_, v: number): string => synIDattr(v, 'name'))
+    # We match "escape" and "symbol" for special items, such as
+    # lispEscapeSpecial or lispBarSymbol.
+    ->match('\cstring\|character\|singlequote\|escape\|symbol\|comment') >= 0
+enddef
+
+# GetSkip {{{2
+if !has('syntax')
+
+  def GetSkip(): func
+    return () => false
+  enddef
+
+else
+
+  def GetSkip(): func
+    if !exists('g:syntax_on') || GetOption('syntax_ignored')
+      return () => false
     else
-      if has("byte_offset") && has("syntax_items") && &smc > 0
-	let stopbyte = max([1, line2byte(".") + col(".") - &smc * 2])
-	let stopline = max([top_viewable, byte2line(stopbyte)])
+      # If evaluating the expression determines that the cursor is
+      # currently in a text with some specific syntax type (like a string
+      # or a comment), then we want searchpairpos() to find a pair within
+      # a text of similar type; i.e. we want to ignore a pair of different
+      # syntax type.
+      if InStringOrComment()
+        return (): bool => !InStringOrComment()
```

Comment:
Shouldn't this be something like:

```
if InString()
    return (): bool => !InString()
elseif InComment()
  return (): bool => !InComment()
else
  return (): bool => InStringOrComment()
endif
```

Are there occasions where one would like parentheses in comments to match with parentheses in strings?

Reply:
Yes, I noticed that too.  The issue is that I don't know what kind of texts the original tokens were supposed to match:

    \ '''v:val =~? "string\\|character\\|singlequote\\|escape\\symbol\\|comment"''))'

What does each of these match:

   - `character`
   - `singlequote`
   - `escape`
   - `symbol`

Some part of a comment?  Some part of a string?  Something else?

Since I don't know, I've just kept the logic of the original plugin which doesn't handle them differently.  With the exception of the test inversion which was wrong.

Reply:
> Are there occasions where one would want parentheses in comments to match with parentheses in strings?

Also, the suggested code wouldn't always work as expected:

    elseif InComment()
      return (): bool => !InComment()

Suppose the cursor is on the opening parenthesis on this comment:

    # (aaa "bbb)" ccc

`InComment()` is true.
`searchpairpos()` finds the closing parenthesis.
On the closing parenthesis, `InComment()` is still true, `!InComment()` is false, the skip expression is false, and `searchpairpos()` does not discard the match.

The end result is that the plugin highlights the closing parenthesis.  But that's not what you wanted; your comment seems to imply that you wanted to ignore a match in a string if the cursor was on a paren in a comment.

What makes things complex is the possibility for syntax items to be nested; like you can have a string highlighted as a string inside a comment.

I'm not saying the current logic can't be improved.  I haven't thought much about it, as I couldn't give more time to this refactoring.  But if you have a specific patch in mind, let me know.